### PR TITLE
Add dropDown React component

### PR DESCRIPTION
**Pull request by JulianRoesner**
# Short Description
To reactify the StepSteuernummer we need to add some components in React. One is the dropDown component. I based the design on this document in Figma and talked to Nadine.

# Changes
- Add dropdown component and relevant icon

# Feedback
- Does this make sense to you?
- Is there anything that I missed while reactifying?

### CODE REVIEW

File: webapp/client/src/stories/FormFieldDropDown.stories.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    title: "Warum muss ich das auswählen?",
+    text: "Die Finanzämter sind nach Bundesländern sortiert.",
+  },
+  values: [
+    ["bw", "Baden-Württemberg"],
+    ["by", "Bayern"],
+    ["hh", "Hamburg"],
+    ["he", "Hessen"],
+  ],
+};
```
Comment (nfelger) [inline webapp/client/src/stories/FormFieldDropDown.stories.js:52]: You can simplify this as:

```suggestion
WithDetails.args = {
  ...Default.args,
  details: {
    title: "Warum muss ich das auswählen?",
    text: "Die Finanzämter sind nach Bundesländern sortiert.",
  },
};
```

File: webapp/client/src/stories/FormFieldDropDown.stories.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  values: [
+    ["bw", "Baden-Württemberg"],
+    ["by", "Bayern"],
+    ["hh", "Hamburg"],
+    ["he", "Hessen"],
+  ],
+  onChangeHandler: (event) => {
+    alert(`Sie leben in ${event.target.selectedOptions[0].text}.`);
+  },
+};
```
Comment (nfelger) [inline webapp/client/src/stories/FormFieldDropDown.stories.js:106]: These can also be simplified as shown above. (This also applies to your other PR - I missed it there…)

File: webapp/client/public/css/base.css [hunk 3]
```diff
@@ -126,6 +126,11 @@ input.form-control.field-error-found {
   min-height: 55px;
 }
 
+.steuerlotse-select:focus {
+  border: 2px solid var(--focus-border-color);
+  box-shadow: 0 0 0 2px var(--focus-color);
+}
+
```
Comment (nfelger) [inline webapp/client/public/css/base.css:133]: Why does this also need to go into the base.css?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  & .steuerlotse-select:focus {
+    border: 2px solid var(--focus-border-color);
+    box-shadow: 0 0 0 2px var(--focus-color);
+  }
+`;
+
+function FormFieldDropDown({
+  fieldName,
+  fieldId,
+  values,
```
Comment (nfelger) [inline webapp/client/src/components/FormFieldDropDown.js:35]: Would this be better called "options"? => clearer mapping to `<option>` tags.

File: webapp/client/src/components/FormFieldDropDown.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  defaultValue,
+  preselectedValue,
```
Comment (nfelger) [inline webapp/client/src/components/FormFieldDropDown.js:37]: What's the difference between preselected and default?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            className="custom-select steuerlotse-select"
+            input_req_err_msg="Bundesland auswählen"
+            name={fieldId}
+            value={preselectedValue}
+            required={required}
+            autoFocus={autofocus || Boolean(errors.length)}
+            onBlur={onChangeHandler}
+            onChange={onChangeHandler}
+          >
+            {defaultValue && <option value="">{defaultValue}</option>}
```
Comment (nfelger) [inline webapp/client/src/components/FormFieldDropDown.js:67]: Why is the value of the default option blank?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        cols: "6",
+      }}
+      hideLabel
+      render={() => (
+        <DropDown>
+          <FieldLabelForSeparatedFields {...{ label, fieldId, details }} />
+          <select
+            id={fieldId}
+            className="custom-select steuerlotse-select"
+            input_req_err_msg="Bundesland auswählen"
```
Comment (nfelger) [inline webapp/client/src/components/FormFieldDropDown.js:59]: Can we implement custom validity with React? Does something roughly like this work?
```suggestion
            onInvalid={(e) => e.target.validity.valueMissing ? "Bundesland auswählen" : ""
```

If it does, we could factor it out into a helper…

**Review by nfelger — CHANGES_REQUESTED** [review]
Looks good, I'm a bit confused on a few things I left comments about, though. This also could use some unit tests.

I especially wonder if this would be a good component to write at least one test that tries to interact with the component using only the keyboard. This library would be good for it: https://testing-library.com/docs/ecosystem-user-event#keyboardtext-options. This is also interesting under "Testing Keyboard Navigation": https://medium.com/moonpigtech/how-we-automate-accessibility-testing-at-moonpig-89a0ae0db0f5

NB: Again, I didn't review CSS nor look at it locally. Let me know if I should!

File: webapp/client/src/components/FormFieldDropDown.js (hunk 5 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:37]: I have renamed defaultValue to `defaultOption`. `defaultOption` Is a default option that is kept empty (like "Bitte auswählen"). The preselectedValue on the other hand would be a selection coming from the backend.

Do you see a better way for naming them or should we add a comment?

File: webapp/client/src/stories/FormFieldDropDown.stories.js (hunk 1 above)
Comment (JulianRoesner) [reply webapp/client/src/stories/FormFieldDropDown.stories.js:52]: Thanks so much. This makes so much more sense

File: webapp/client/public/css/base.css (hunk 3 above)
Comment (JulianRoesner) [reply webapp/client/public/css/base.css:133]: Because so far the focus of that component was missing and this should add this not only for the react component but the old component as well.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 6 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:67]: I could not find another way to have an empty value for the defaultValue. If I omit the value, the react component will have the name of the default option as the value (e.g. "Bitte auswählen"). However, we do not want that to have a correct value.

Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:67]: That said, because of your comment about the custom validity, I am checking if it would make sense to make that a controlled input

File: webapp/client/src/components/FormFieldDropDown.js (hunk 5 above)
Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Thanks! I understand `defaultOption` now. For prefilled data that comes from the backend, we use `value` in our other fields. `<select>`s are different though... Maybe a more consistent name could be `selectedOption`? On the other hand, React uses `defaultValue` to mark an option as selected on an uncontrolled select, so maybe `value` does work?

Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Wait: `selectedValue`, not `selectedOption`, because `defaultValue` takes the value.

Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: This probably just confused you more! :)

File: webapp/client/src/components/FormFieldDropDown.js (hunk 6 above)
Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Aha, with the changed names, this makes perfect sense. I was just confused.

Why does the custom validity lead you in the direction of a controlled input?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 8]
```diff
@@ -52,21 +43,21 @@ function FormFieldDropDown({
...
             autoFocus={autofocus || Boolean(errors.length)}
             onBlur={onChangeHandler}
             onChange={onChangeHandler}
           >
-            {defaultValue && <option value="">{defaultValue}</option>}
-            {values.map((value) => (
-              <option value={value[0]}>{value[1]}</option>
+            {defaultOption && <option value="">{defaultOption}</option>}
+            {options.map((option) => (
+              <option value={option[0]}>{option[1]}</option>
```
Comment (nfelger) [inline webapp/client/src/components/FormFieldDropDown.js:60]: Hm, I think this would be more readable if we made the `option`s into objects instead of an arrays? Something like:
```js
{
  value: "by",
  displayName: "Bayern"
}
```

File: webapp/client/src/components/FormFieldDropDown.js (hunk 6 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js]: I thought that the problem might be that the react component does not set the value correctly and the validation might not work because of that. That was the reason that I thought we might need a controlled input for that.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 7 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:50]: Thank you so much for pointing that out. I just copied that over from the component in the backend. 

I had another look into this and I think that we could get rid of that completely. We do not use that because we set `novalidate` on all our forms which prevents client validation. We just use the validation messages from the backend. The `input_req_err_msg` has been dead code as far as I can tell. In the future we might want to move that also to the frontend. However, I think it might make more sense to use one of the libraries that we looked into for that.

That said, while looking into it I discovered that it would work if you do something like:
```react
onInvalid={(e) => e.target.setCustomValidity("Bundesland auswählen")
```

TLDR: We can get rid of that because the backend will take care of this.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 5 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js]: So,`defaultOption` and `selectedValue`?

File: webapp/client/src/components/FormFieldDropDown.js (hunk 8 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js]: Great idea 👍

File: webapp/client/src/components/FormFieldDropDown.js (hunk 6 above)
Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Interesting, ok. Guess we won't find out this time around :)

File: webapp/client/src/components/FormFieldDropDown.js (hunk 7 above)
Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Awesome. Deleted code is the best code :)

**Review by nfelger — APPROVED** [review]

File: webapp/client/src/components/FormFieldDropDown.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      render={() => (
+        <DropDown>
+          <FieldLabel {...{ label, fieldId, details }} />
+          <select
+            id={fieldId}
+            className="custom-select steuerlotse-select"
+            name={fieldId}
+            defaultValue={selectedValue}
+            required={required}
+            autoFocus={Boolean(errors.length)}
```
Comment (frederike-ramin) [inline webapp/client/src/components/FormFieldDropDown.js:52]: I think thos should also have a `autofocus` parameter that gets tested here? Or is it a concious decision to leave the `autofocus` out here because we want to get rid of it in the future anyway?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  fieldId,
+  options,
+  defaultOption,
+  selectedValue,
+  required,
+  label,
+  details,
+  errors,
+  onChangeHandler,
+}) {
```
Comment (frederike-ramin) [inline webapp/client/src/components/FormFieldDropDown.js:34]: Why did you decide to pass the `onChangeHandler` here? Do you expect it to be different for different drop downs? Otherwise we could just declare it for this component, riht?

File: webapp/client/src/components/FormFieldDropDown.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            required={required}
+            autoFocus={Boolean(errors.length)}
+            onBlur={onChangeHandler}
+            onChange={onChangeHandler}
+          >
+            {defaultOption && (
+              <option value="" key={fieldId + defaultOption}>
+                {defaultOption}
+              </option>
+            )}
```
Comment (frederike-ramin) [inline webapp/client/src/components/FormFieldDropDown.js:60]: Shouldn't the default option just be set in the backend to the wtforms field? Right now it is in some cases (e.g. Bundesland) while for the tax offices it's set in JavaScript.

File: webapp/client/src/components/FormFieldDropDown.test.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      ],
+      onChangeHandler: onChangeHandler,
+    };
+    render(<FormFieldDropDown {...props} />);
+  });
+
+  describe("When second option selected", () => {
+    beforeEach(() => {
+      userEvent.selectOptions(screen.getByRole("combobox"), ["B"]);
+    });
```
Comment (frederike-ramin) [inline webapp/client/src/components/FormFieldDropDown.test.js:31]: Do we need a test for changing the value?

**Review by frederike-ramin — COMMENTED** [review]
The component looks great. 👌 
I just have quite some questions and suggestions if you wanna have a look at that. :)

File: webapp/client/src/components/FormFieldDropDown.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    background-image: url(${selectIcon});
+    background-repeat: no-repeat;
+    background-size: 0.75rem;
+    min-height: 55px;
+  }
+
+  & .steuerlotse-select:focus {
+    border: 2px solid var(--focus-border-color);
+    box-shadow: 0 0 0 2px var(--focus-color);
+  }
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:20]: Interesting. I think you are right that we do not need the `&`here. 

However, my understanding of the code is different (based on [this part in the documentation](https://styled-components.com/docs/basics#pseudoelements-pseudoselectors-and-nesting)):
- The ampersand refers back to the main styled component (DropDown)
- If we want to refer only to classes added to the styled component we would do `&.class-name`
- `& .class-name` would refer to children of the styled component with that class. This is exactly what we want. However, we would achieve the same with the ampersand.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 9 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:52]: That was basically my thinking

File: webapp/client/src/components/FormFieldDropDown.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/client/src/components/FormFieldDropDown.js:20]: But isn't the css of a `styled` component already just scoped to that component? Making every class declaration within it local to that single component (and its children)?

File: webapp/client/src/components/FormFieldDropDown.js (hunk 10 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:34]: The onChangeHandler is needed to change the state in the parent component. Therefore, it should be adaptable and we are not able to set it in this component because we need it to change the state in the parent component.
That is why we have to pass it

File: webapp/client/src/components/FormFieldDropDown.js (hunk 11 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:60]: I think I do not really get what you mean? We need to set the defaultOption in the React component if it is coming from the backend or if it is always the same for all components.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 10 above)
Comment (frederike-ramin) [reply webapp/client/src/components/FormFieldDropDown.js:34]: Ah..right, I see, thanks! Do we need it onBlur and onChange though? Wouldn't onChange be enough? Probably just asking dumb questions here. :D

File: webapp/client/src/components/FormFieldDropDown.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    <FormFieldScaffolding
+      {...{
+        fieldName,
+        errors,
+        cols: "6",
+      }}
+      hideLabel
+      render={() => (
+        <DropDown>
+          <FieldLabel {...{ label, fieldId, details }} />
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:45]: I just copied that part from our other components. However, I think that is a great idea and makes much more sense. Thank you 👍

File: webapp/client/src/components/FormFieldDropDown.test.js (hunk 12 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.test.js:31]: Great idea 👍

File: webapp/client/src/components/FormFieldDropDown.js (hunk 11 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:60]: Okay, I think I understood you correctly then :) I think there are two aspects of that problem:
1. We need to tell the frontend that we have a default option. So far, that could be done by the backend because we used WTForms for the creation of the HTML. When using this React component it needs to take care of setting the default option. That is why this part is here.
2. The integration aspect. We could change it here to not be adaptable and always say "Bitte auswählen" or at least be the default option. I think that this makes a lot of sense and we should change that. However, I would leave that aspect to the integration PR.

I hope that this makes it a little bit clearer. Otherwise, let's talk ;)

File: webapp/client/src/components/FormFieldDropDown.js (hunk 10 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:34]: Not at all a dumb question. When I was developing this feature we had an older version of `eslint-plugin-jsx-a11y` in use which enforced the following rule: [jsx-a11y/no-onchange](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-onchange.md) - use onBlur on select events because some browser do not support onChange correctly for selects

That rule has been deprecated and we now use a newer version of the plugin that does not enforce the rule. Therefore, we could get rid of onBlur, now.

File: webapp/client/src/components/FormFieldDropDown.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js:20]: That's why:
>  I think you are right that we do not need the `&` here.

Sorry, for the nit-picking explanation that was basically saying: There is a difference between `&.class` and `& .class`

File: webapp/client/src/components/FormFieldDropDown.js (hunk 11 above)
Comment (frederike-ramin) [reply webapp/client/src/components/FormFieldDropDown.js]: Oh god, I'm a bit afraid to ask..but why?
> We need to tell the frontend that we have a default option.
Isn't that just another option? Or is it special to the frontend in any way?

Comment (frederike-ramin) [reply webapp/client/src/components/FormFieldDropDown.js]: I think this didn't become clear in my comments so far: I'm not at all convinced we *should* set it in the backend. Just that we should do it consistently.

Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: I'm with Frede here. Picking what default option to show feels like a presentation thing, not a data thing, and therefore belongs into the frontend.

Comment (nfelger) [reply webapp/client/src/components/FormFieldDropDown.js]: Actually, maybe that's what you're already suggesting @JulianRoesner? (Definitely something for the integration PR though, so let's maybe have that conversation then…)

Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.js]: > Actually, maybe that's what you're already suggesting @JulianRoesner? 

Yes, that's what I meant.

File: webapp/client/src/components/FormFieldDropDown.test.js (hunk 12 above)
Comment (JulianRoesner) [reply webapp/client/src/components/FormFieldDropDown.test.js]: This lead me to a rabbit hole because the tests for interacting with the Dropdown component via keyboard did not really work as expected. I was not able to get the tests running with the user event library. I have opened an issue in their repository and I will come back to this when ever there is something to follow up and add correct tests for the interaction via keyboard
